### PR TITLE
Bloody footprints!

### PR DESCRIPTION
**Pull request by RemieRichards**
- Human blood, Xeno blood and Oil (Borg blood) now apply blood to shoes when you walk over them
- Human blood, Xeno blood and Oil can now make footprints when somebody walks, capable of storing 8 different kinds of footprint per blood type (Coming in east, Leaving east, coming in west, leaving west, etc.)
- Blood types stack and also mix their colour if stacked
- Human blood will apply the bloody shoe overlay to bloody shoes
- Footprints fade as you use up the blood on the shoe

GiffyBlood

### CODE REVIEW

**Comment by Xhuis** [comment]
rip janitors

**Comment by WJohn** [comment]
I dunno about this, last time we had this a couple of years ago it lagged the server to shit because there were always thousands and thousands of footprints everywhere, stacking up endlessly. I hope you've addressed that problem.

**Comment by Lumipharon** [comment]
Remember when Pap or whoever made dirt spread by walking around?
It was fucking awful and resulted with permanent ghetto station 24/7, even when it got toned down.

That said, I think it could be cool if done in a way that doesn't result in the above. It would also require that blood dries (quickly) and you don't make bloody prints for a long time after.

**Comment by RemieRichards** [comment]
@WJohn 
1. Images are cached, there are, at most, 8(4*enter+exit) * 3(Human,Alien,Oil), so 24 Images in existence.
2. Updating the icon state of the footprint objects uses bitflags, there's really no faster way to do that besides a little bit more bitflag magic.
3. There is only 1 footprint "object" per turf/tile (so identical to blood trails) so the amount of footprint objects is equal to the turfs they cover, meaning no rightclick lag or anything.

This essentially means that it's cheap as chips.

@Lumipharon one step on a blood splatter (not an old one) is 50 "bloodiness", this will get you about 8 ish (I can't remember, it's not very far) footprints, and walking back over the same footprints doesn't increase the blood or duplicate it or anything, it'll causes messes during fights or when someone walks over blood, but it won't track ALL THE WAY ACROSS THE STATION, so it's fine.
And the amount of blood loss per step (which effectively controls the number of footprints one crossing of blood produces) is 100% tweakable.

**Comment by WJohn** [comment]
Very nice, then :+1:

**Comment by Lumipharon** [comment]
Just keep in mind if this works with blood trails from dragging bodies, we WILL see khorne station all the time.

**Comment by kevinz000** [comment]
This reminds me of bay :(

**Comment by WJohn** [comment]
I think the sprites are actually straight FROM bay.

**Comment by RemieRichards** [comment]
@WJohn depends where they came from originally, as these are the original 4 year old sprites.
If footprints from then came from a bay port, then perhaps so, but bay don't currenty use these sprites to my knowledge.

File: code/game/objects/effects/decals/Cleanable/humans.dm [hunk 1]
```diff
@@ -104,4 +106,56 @@
...
 	..()
 	spawn(1)
-		drips |= icon_state
\ No newline at end of file
+		drips |= icon_state
+	bloodiness = rand(5,15)
+
+
+//BLOODY FOOTPRINTS
+/obj/effect/decal/cleanable/blood/footprints
```
Comment (optimumtact) [inline code/game/objects/effects/decals/Cleanable/humans.dm]: are these pool able?

Are all the cleanable pool able.

**Comment by SconesC** [comment]
This is going to get really old really fast

**Comment by Razharas** [comment]
I remember in the past a lot of blood-spread diseases on a dirty station were lagging

**Comment by RemieRichards** [comment]
That has nothing to do with this at all @Razharas

**Comment by Giacom** [comment]
Looking at the gif, it would probably look better if you only leave behind footprints and they aren't created from thinair as you're walking (gliding) to the tile.

**Comment by Dorsisdwarf** [comment]
Agreed.

Also this will be excellent when a murderboning ling with digital camo goes all invisible-man on the AI

Do invisible people leave tracks?
Do mechs leave tracks?
Do people using the chameleon projector leave tracks?

**Comment by Miauw62** [comment]
blood should thin out when people walk through it so one tile of blood cant make infinite tracks

**Comment by ike709** [comment]
^ What Miauw said is probably the one thing that could make this even better

**Comment by Aranclanos** [comment]
another check for each step that mobs do?

**Comment by Jalleo** [comment]
@Miauw62 you dont read
And the amount of blood loss per step (which effectively controls the number of footprints one crossing of blood produces) is 100% tweakable. With one step being 50 bloodiness, a fresh blood splatter can only contain 100 "bloodiness", so you will at most, be able to make 2 small trails per pool of blood (again, totally tweakable)

@remierichards firstly please make sure that info is bolded at top of the PR
Second can you have the pool of blood convert to dried blood once it ran out of blood?
Third does trails of blood cause smaller trails of this or none?

**Comment by RemieRichards** [comment]
@Miauw62 as Jalleo quoted, I've already explained pools of blood fade out as you take "bloodiness" from them, with bloodiness ranging from 0-100, with the blood decal deleting itself when it reaches 0 or lower.

Trails don't produce footprints, and footprints don't produce more footprints (except on the same tile if you walk around it in all directions, I'm counting this as "spreading" the blood, but not picking it up on your shoes)

**Comment by Razharas** [comment]
What about shoeless walking?

**Comment by Alek2ander** [comment]
Slaughter demon buff. Oh joy.

**Comment by Supermichael777** [comment]
actualy checking for shoes would alowfor uneque footprints for shoe types. but thats beyond the scope of this pr

**Comment by Razharas** [comment]
Ok ill pretend taking off shoes to not leave the prints is intended gameplay mechanic to balance something something

**Comment by Durkel** [comment]
It's been four hours and this is already fucking awful on live. Stop adding this crap, jesus.

**Comment by duncathan** [comment]
What's wrong with it?

**Pull request merged**